introduction_to_applying_machine_learning/gluon_recommender_system/recommender.py

```diff
@@ -0,0 +1,254 @@
+import logging
+import json
+import time
+import os
+import mxnet as mx
+from mxnet import gluon, nd, ndarray
+from mxnet.metric import MSE
+import numpy as np
+
+import pip
+pip.main(['install', 'pandas'])
+import pandas as pd
+
+logging.basicConfig(level=logging.DEBUG)
+
+#########
+# Globals
+#########
+
+batch_size = 40
+
+
+##########
+# Training
+##########
+
+def train(channel_input_dirs, hyperparameters, hosts, num_gpus, **kwargs):
+    
+    # get data
+    training_dir = channel_input_dirs['train']
+    train_iter, test_iter, customer_index, product_index = prepare_train_data(training_dir)
+    
+    # get hyperparameters
+    num_embeddings = hyperparameters.get('num_embeddings', 64)
+    opt = hyperparameters.get('opt', 'sgd')
+    lr = hyperparameters.get('lr', 0.02)
+    momentum = hyperparameters.get('momentum', 0.9)
+    wd = hyperparameters.get('wd', 0.)
+    epochs = hyperparameters.get('epochs', 5)
+
+    # define net
+    ctx = mx.gpu()
+
+    net = MFBlock(max_users=customer_index.shape[0], 
+                  max_items=product_index.shape[0],
+                  num_emb=num_embeddings,
+                  dropout_p=0.)
+    net.collect_params()
+    
+    net.collect_params().initialize(mx.init.Xavier(magnitude=2.24),
+                                    ctx=ctx,
+                                    force_reinit=True)
+
+    trainer = gluon.Trainer(net.collect_params(),
+                            opt,
+                            {'learning_rate': lr,
+                             'wd': wd,
+                             'momentum': momentum})
+    
+    # execute
+    trained_net = execute(train_iter, test_iter, net, trainer, epochs, ctx)
+    
+    return trained_net, customer_index, product_index
+
+
+class MFBlock(gluon.Block):
+    def __init__(self, max_users, max_items, num_emb, dropout_p=0.5):
+        super(MFBlock, self).__init__()
+        
+        self.max_users = max_users
+        self.max_items = max_items
+        self.dropout_p = dropout_p
+        self.num_emb = num_emb
+        
+        with self.name_scope():
+            self.user_embeddings = gluon.nn.Embedding(max_users, num_emb)
+            self.item_embeddings = gluon.nn.Embedding(max_items, num_emb)
+            self.dropout = gluon.nn.Dropout(dropout_p)
+            self.dense = gluon.nn.Dense(num_emb, activation='relu')
+            
+    def forward(self, users, items):
```

Comment:
Please use hybrid_forward. The operation here is simple and should be easily hybridized.